wsd/DocumentBroker.cpp

```diff
@@ -1457,6 +1495,388 @@ DocumentBroker::updateSessionWithWopiInfo(const std::shared_ptr<ClientSession>&
+
+            Poco::Path destDir(_presetsPath, "xcu");
+            Poco::File(destDir).createDirectories();
+            std::string fileName = Poco::Path(destDir, "config.xcu").toString();
+
+            queries.emplace_back(uri, stamp, fileName);
+        }
+    }
+public:
+    PresetsInstallTask(SocketPoll& poll, const std::string& configId,
+                       const std::string& presetsPath,
+                       std::function<void(bool)> installFinishedCB)
+        : _poll(poll)
+        , _reportedStatus(false)
+        , _overallSuccess(true)
+        , _idCount(0)
+        , _configId(configId)
+        , _presetsPath(presetsPath)
+    {
+        appendCallback(installFinishedCB);
+    }
+
+    bool empty() const
+    {
+        return _installingPresets.empty();
+    }
+
+    void appendCallback(std::function<void(bool)> installFinishedCB)
+    {
+        _installFinishedCBs.emplace_back(installFinishedCB);
+    }
+
+    void install(Poco::JSON::Object::Ptr settings)
+    {
+        std::vector<CacheQuery> presets;
+        if (!settings)
+            _overallSuccess = false;
+        else
+        {
+            addGroup(settings, "autotext", presets);
+            addGroup(settings, "wordbook", presets);
+            addXcu(settings, presets);
+        }
+
+        Cache::supplyConfigFiles(_configId, presets);
+
+        // If there are no presets to fetch then we can respond now, otherwise
+        // that happens when the last preset is installed.
+        if (!presets.empty())
+        {
+            LOG_INF("Async fetch of presets for " << _configId << " launched");
+            for (const auto& preset : presets)
+                asyncInstall(preset._uri, preset._stamp, preset._dest);
+        }
+        else
+        {
+            LOG_INF("Fetch of presets for " << _configId << " completed immediately. Success: " << _overallSuccess);
+            completed();
+        }
+    }
+};
+
+void DocumentBroker::asyncInstallPresets(const std::shared_ptr<ClientSession>& session,
+                                         const std::string& userSettingsUri,
+                                         const std::string& presetsPath)
+{
+    auto installFinishedCB =
+        [selfWeak = weak_from_this(), this, session, userSettingsUri, presetsPath](bool success)
+    {
+        std::shared_ptr<DocumentBroker> selfLifecycle = selfWeak.lock();
+        if (!selfLifecycle)
+            return;
+
+        if (success)
+        {
+            std::string searchDir = presetsPath;
+            searchDir.append("wordbook");
+            const auto fileNames = FileUtil::getDirEntries(searchDir);
+            for (auto& fileName : fileNames)
+            {
+                std::string filePath = searchDir;
+                filePath.append("/");
+                filePath.append(fileName);
+                _presetTimestamp[fileName] = FileUtil::getLastModificationTimestamp(filePath);
```

Comment:
Should we skip inserting/updating the map if `getLastModificationTimestamp()` fails (file doesn't exist or some other error) ?